Case-let doesn't look through ticks

While investigating #199 I noticed that the example stopped compiling at all, even with the suggested raised inline limit. This is due to `caseLet` not looking though ticks.

Compiling:
```haskell
module Test where

import Clash.Prelude

f :: BitVector 5 -> Maybe (Bit -> Bit -> Bit)
f i = case slice d1 d0 i of
    2 -> op
    3 -> op
    _ -> Nothing
  where op :: Maybe (Bit -> Bit -> Bit)
        op = case (slice d4 d2 i, b) of
            (0, 0) -> Just (+)
            _ -> Nothing
          where b :: Bit
                b = case i ! 0 of
                    0 -> 0
                    1 -> 0

topEntity :: BitVector 5 -> Bit -> Bit -> Maybe Bit
topEntity = (sequenceA .) . sequenceA . f
```

gives:

```
GHC: Parsing and optimising modules took: 0.564s
GHC: Loading external modules from interface files took: 0.121s
GHC: Parsing annotations took: 0.000s
Clash: Parsing and compiling primitives took 2.540s
GHC+Clash: Loading modules cumulatively took 3.416s
Clash: Compiling Test.topEntity
Clash: Applied 75 transformations
Clash: Normalisation took 0.003s

Test.hs:20:1: error:
    Clash.Netlist(278): Can't translate non-tycon type: Clash.Sized.Internal.BitVector.Bit8214565720323789615
-> Clash.Sized.Internal.BitVector.Bit8214565720323789615
-> Clash.Sized.Internal.BitVector.Bit8214565720323789615
    NB: The source location of the error is not exact, only indicative, as it is acquired after optimisations.
    The actual location of the error can be in a function that is inlined.
    To prevent inlining of those functions, annotate them with a NOINLINE pragma.
   |
20 | topEntity = (sequenceA .) . sequenceA . f
   | ^^^^^^^^^
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -75,6 +75,7 @@ Closed
   * [#347](https://github.com/clash-lang/clash-compiler/issues/347): topEntity synthesis fails due to insufficient type-level normalisation
   * [#626](https://github.com/clash-lang/clash-compiler/issues/626): Missing Clash.Explicit.Prelude definitions
   * [#960](https://github.com/clash-lang/clash-compiler/issues/626): Blackbox Error Caused by Simple map
+  * [#1012](https://github.com/clash-lang/clash-compiler/issues/1012): Case-let doesn't look through ticks
 
 * Fixes without issue reports:
   * Fix bug in `rnfX` defined for `Down` ([baef30e](https://github.com/clash-lang/clash-compiler/commit/baef30eae03dc02ba847ffbb8fae7f365c5287c2))

--- a/clash-lib/src/Clash/Normalize/Transformations.hs
+++ b/clash-lib/src/Clash/Normalize/Transformations.hs
@@ -258,8 +258,8 @@ nonRepSpec _ e = return e
 
 -- | Lift the let-bindings out of the subject of a Case-decomposition
 caseLet :: HasCallStack => NormRewrite
-caseLet _ (Case (Letrec xes e) ty alts) =
-  changed (Letrec xes (Case e ty alts))
+caseLet _ (Case (collectTicks -> (Letrec xes e,ticks)) ty alts) =
+  changed (Letrec (map (second (`mkTicks` ticks)) xes) (Case (mkTicks e ticks) ty alts))
 
 caseLet _ e = return e
 

--- a/tests/shouldwork/Basic/T1012.hs
+++ b/tests/shouldwork/Basic/T1012.hs
@@ -1,0 +1,20 @@
+module T1012 where
+
+import Clash.Prelude
+
+f :: BitVector 5 -> Maybe (Bit -> Bit -> Bit)
+f i = case slice d1 d0 i of
+    2 -> op
+    3 -> op
+    _ -> Nothing
+  where op :: Maybe (Bit -> Bit -> Bit)
+        op = case (slice d4 d2 i, b) of
+            (0, 0) -> Just (+)
+            _ -> Nothing
+          where b :: Bit
+                b = case i ! 0 of
+                    0 -> 0
+                    1 -> 0
+
+topEntity :: BitVector 5 -> Bit -> Bit -> Maybe Bit
+topEntity = (sequenceA .) . sequenceA . f

--- a/testsuite/Main.hs
+++ b/testsuite/Main.hs
@@ -137,6 +137,7 @@ runClashTest = defaultMain $ clashTestRoot
         , runTest "Shift" def{hdlSim=False}
         , runTest "SimpleConstructor" def{hdlSim=False}
         , runTest "T347" def{hdlSim=False}
+        , runTest "T1012" def{hdlSim=False}
         , runTest "TagToEnum" def{hdlSim=False}
         , runTest "TestIndex" def{hdlSim=False}
         , runTest "Time" def

